File: ios/ClaviKeyboard/ClaviKeyboardExtension/TranslationEngine.swift
import Foundation

/**
 * Two-tier translation engine (iOS mirror of TranslationEngine.kt).
 *
 * Tier 1: Static TSV word dictionary bundled in Resources/translate/{src}-{tgt}.tsv
 *         Zero download, word-level, instant.
 * Tier 2: Apple Translation framework (iOS 17+), sentence-level.
 *         Falls back to Tier 1 if iOS < 17 or translation fails.
 */
struct TranslationEngine {

    struct TranslationSuggestion {
        let original: String    // phrase that was translated
        let translated: String  // translation result
        let sourceLang: String
        let targetLang: String
    }

    let sourceLang: String
    let targetLang: String

    private let dictionary: [String: String]

    init(sourceLang: String, targetLang: String) {
        self.sourceLang = sourceLang
        self.targetLang = targetLang
        self.dictionary = TranslationEngine.loadDict(src: sourceLang, tgt: targetLang)
    }

    /**
     * Async translate. iOS 17+: Apple Translation framework (sentence-level).
     * Older iOS or failure: falls back to word-level TSV lookup.
     * Calls [completion] on the main thread.
     */
    func translate(_ phraseBeforeCursor: String, completion: @escaping (TranslationSuggestion?) -> Void) {
        guard let phrase = extractPhrase(phraseBeforeCursor) else {
            completion(nil)
            return
        }

        if #available(iOS 17, *) {
            Task {
                do {
                    let result = try await appleTranslate(phrase)
                    await MainActor.run {
                        completion(TranslationSuggestion(
                            original: phrase, translated: result,
                            sourceLang: sourceLang, targetLang: targetLang))
                    }
                } catch {
                    // Fall back to Tier 1
                    let word = lastWord(of: phrase)
                    let result = dictionary[word.lowercased()]
                    await MainActor.run {
                        completion(result.map {
                            TranslationSuggestion(original: word, translated: $0,
                                                  sourceLang: sourceLang, targetLang: targetLang)
                        })
                    }
                }
            }
        } else {
            // iOS < 17: word-level TSV only
            let word = lastWord(of: phrase)
            if let result = dictionary[word.lowercased()] {
                completion(TranslationSuggestion(original: word, translated: result,
                                                 sourceLang: sourceLang, targetLang: targetLang))
            } else {
                completion(nil)
            }
        }
    }

    // MARK: - Apple Translation (iOS 17+)

    @available(iOS 17, *)
    private func appleTranslate(_ text: String) async throws -> String {
        let src = Locale.Language(identifier: sourceLang)
        let tgt = Locale.Language(identifier: targetLang)
        let config = TranslationSession.Configuration(source: src, target: tgt)
        let session = TranslationSession(configuration: config)
        let response = try await session.translate(text)
        return response.targetText
    }

    // MARK: - Phrase extraction

    private func extractPhrase(_ text: String) -> String? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        let sentenceEnders = CharacterSet(charactersIn: ".!?\n")
        let phrase: String
        if let range = trimmed.rangeOfCharacter(from: sentenceEnders, options: .backwards) {
            phrase = String(trimmed[trimmed.index(after: range.lowerBound)...])
                .trimmingCharacters(in: .whitespaces)
        } else {
            phrase = trimmed
        }
        return phrase.count > 2 ? phrase : nil
    }

    private func lastWord(of phrase: String) -> String {
        phrase.trimmingCharacters(in: .whitespaces)
            .components(separatedBy: " ")
            .last(where: { !$0.isEmpty }) ?? phrase
    }

    // MARK: - TSV dictionary loader

    private static func loadDict(src: String, tgt: String) -> [String: String] {
        guard let url = Bundle.main.url(forResource: "\(src)-\(tgt)", withExtension: "tsv",
                                        subdirectory: "translate"),
              let content = try? String(contentsOf: url, encoding: .utf8) else { return [:] }
        var dict: [String: String] = [:]
        for line in content.components(separatedBy: "\n") {
            guard !line.hasPrefix("#"), !line.isEmpty,
                  let tabIdx = line.firstIndex(of: "\t") else { continue }
            let key = String(line[..<tabIdx])
            let value = String(line[line.index(after: tabIdx)...])
            dict[key] = value
        }
        return dict
    }
}

File: android/app/src/main/java/com/clavi/keyboard/TranslationEngine.kt
package com.clavi.keyboard

import android.content.Context
import com.google.android.gms.tasks.Task
import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.nl.translate.TranslateLanguage
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.TranslatorOptions
import java.io.IOException

/**
 * Two-tier translation engine.
 *
 * Tier 1: Static TSV word dictionary bundled in assets/translate/{src}-{tgt}.tsv
 *         Zero download, word-level, instant.
 * Tier 2: ML Kit on-device translator (sentence-level, ~30MB model download on WiFi).
 *         Falls back to Tier 1 if model not yet downloaded or translation fails.
 *
 * Usage:
 *   val engine = TranslationEngine("en", "uk", context)
 *   engine.translate("Hello how are you") { suggestion -> ... }
 */
class TranslationEngine(
    val sourceLang: String,
    val targetLang: String,
    private val context: Context,
) {

    data class TranslationSuggestion(
        val original: String,    // phrase that was translated
        val translated: String,  // translation result
        val sourceLang: String,
        val targetLang: String,
    )

    // Tier 1: static TSV dictionary, lazy-loaded
    private val dictionary: Map<String, String> by lazy { loadDict() }

    // Tier 2: ML Kit on-device translator, lazy-created
    private val mlTranslator by lazy {
        val src = TranslateLanguage.fromLanguageTag(sourceLang) ?: TranslateLanguage.ENGLISH
        val tgt = TranslateLanguage.fromLanguageTag(targetLang) ?: TranslateLanguage.UKRAINIAN
        Translation.getClient(
            TranslatorOptions.Builder()
                .setSourceLanguage(src)
                .setTargetLanguage(tgt)
                .build()
        )
    }

    /**
     * Async translate. Tries ML Kit first; falls back to TSV word lookup on failure.
     * Calls [callback] on the calling thread (usually main thread via Task listeners).
     * Returns null if there's nothing worth translating in the given text.
     */
    fun translate(phraseBeforeCursor: String, callback: (TranslationSuggestion?) -> Unit) {
        val phrase = extractPhrase(phraseBeforeCursor) ?: return callback(null)
        mlTranslator.translate(phrase)
            .addOnSuccessListener { result ->
                callback(TranslationSuggestion(phrase, result, sourceLang, targetLang))
            }
            .addOnFailureListener {
                // Tier 1 fallback: look up the last word in the static dictionary
                val lastWord = phrase.trim().substringAfterLast(' ').ifEmpty { phrase.trim() }
                val result = dictionary[lastWord.lowercase()]
                callback(result?.let { TranslationSuggestion(lastWord, it, sourceLang, targetLang) })
            }
    }

    /**
     * Pre-fetch the ML Kit model over WiFi. Call from Settings "Download model" button.
     * [onDone] is called with true on success, false on failure.
     */
    fun downloadModelIfNeeded(onDone: (Boolean) -> Unit) {
        val conditions = DownloadConditions.Builder().requireWifi().build()
        mlTranslator.downloadModelIfNeeded(conditions)
            .addOnSuccessListener { onDone(true) }
            .addOnFailureListener { onDone(false) }
    }

    // ── Phrase extraction ──

    /**
     * Extract the phrase to translate from text before cursor.
     * Returns the text after the last sentence-ending punctuation (or the whole text),
     * trimmed. Returns null if too short to bother translating.
     */
    private fun extractPhrase(text: String): String? {
        val trimmed = text.trimEnd()
        if (trimmed.isEmpty()) return null
        val sentenceStart = trimmed.lastIndexOfAny(charArrayOf('.', '!', '?', '\n'))
        val phrase = if (sentenceStart >= 0) trimmed.substring(sentenceStart + 1).trim()
                     else trimmed
        return phrase.takeIf { it.length > 2 }
    }

    // ── TSV dictionary loader ──

    private fun loadDict(): Map<String, String> {
        return try {
            context.assets.open("translate/$sourceLang-$targetLang.tsv")
                .bufferedReader()
                .lineSequence()
                .filter { it.isNotBlank() && !it.startsWith('#') }
                .mapNotNull { line ->
                    val tab = line.indexOf('\t')
                    if (tab < 0) null else line.substring(0, tab) to line.substring(tab + 1)
                }
                .toMap()
        } catch (_: IOException) {
            emptyMap()
        }
    }
}
